src/overlays/actors/ovl_Obj_Boat/z_obj_boat.c

```diff
@@ -28,31 +30,162 @@ const ActorInit Obj_Boat_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/ObjBoat_Init.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/ObjBoat_Destroy.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/func_80B9B124.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/ObjBoat_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/func_80B9B428.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Boat/ObjBoat_Draw.s")
+s16 func_80B9AF50(ObjBoat* this, Vec3f* arg0) {
+    s16 yaw;
+    Vec3s* temp = &this->unk_164[(s32)this->unk_15C];
+
+    Math_Vec3s_ToVec3f(arg0, &temp[this->unk_15D]);
+    yaw = Math_Vec3f_Yaw(&this->dyna.actor.world.pos, arg0);
+
+    return ((this->unk_15D > 0) ? yaw : (yaw + 0x8000));
+}
+
+void ObjBoat_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad[2];
+    Path* path;
+    ObjBoat* this = THIS;
+    Vec3f sp24;
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyActor_Init(&this->dyna, 3);
+    DynaPolyActor_LoadMesh(globalCtx, &this->dyna, &object_kaizoku_obj_Colheader_009A88);
+    if (thisx->params < 0) {
+        this->dyna.actor.update = func_80B9B428;
+    } else {
+        path = &globalCtx->setupPathList[OBJBOAT_GET_PATH(thisx)];
+        this->unk_163 = path->count - 1;
+        this->unk_164 = Lib_SegmentedToVirtual(path->points);
+        this->unk_15D = 1;
+        this->dyna.actor.world.pos.x = this->unk_164[this->unk_15C].x;
+        this->dyna.actor.world.pos.z = this->unk_164[this->unk_15C].z;
+        this->dyna.actor.shape.rot.y = func_80B9AF50(this, &sp24);
+        this->dyna.actor.world.rot.y = this->dyna.actor.shape.rot.y;
+        this->unk_15D = -this->unk_15D;
+    }
+}
+
+void ObjBoat_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ObjBoat* this = THIS;
+
+    DynaPoly_DeleteBgActor(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+}
+
+void func_80B9B124(ObjBoat* this) {
+    this->unk_160 += 1000;
+    this->dyna.actor.world.pos.y = Math_SinS(this->unk_160) + this->dyna.actor.home.pos.y;
+    this->dyna.actor.shape.rot.x = (Math_SinS(this->unk_160) * 100.0f);
+    this->dyna.actor.shape.rot.z = (Math_SinS((s16)(this->unk_160 * 2)) * 50.0f);
+}
+
+void ObjBoat_Update(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    ObjBoat* this = THIS;
+    Player* player = GET_PLAYER(globalCtx);
+    s32 temp = DynaPolyActor_IsInRidingMovingState(&this->dyna);
+    f32 sp3C = 0.0f;
+    s16 sp3A = this->dyna.actor.shape.rot.y;
+    Vec3f sp2C;
+
+    if ((temp != 0) || ((DynaPolyActor_IsInRidingFallingState(&this->dyna)))) {
+        if ((this->unk_15F == 0) && (OBJBOAT_GET_4000(thisx) || ((temp != 0) && (this->unk_15C == this->unk_15E)))) {
+            this->unk_15D = -this->unk_15D;
+            if (this->unk_15D > 0) {
+                this->unk_15E = this->unk_163;
+            } else {
+                this->unk_15E = 0;
+            }
+            this->unk_15F = 0x3C;
```

Comment:
```suggestion
            this->unk_15F = 60;
```
since this function at least seems to be treating it as a timer.